linked_list/linked_list.py

```diff
@@ -9,71 +9,159 @@ def __init__(self, value, next_node = None):
 # Defines the singly linked list
 class LinkedList:
     def __init__(self):
-      self.head = None # keep the head private. Not accessible outside this class
+        self.head = None # keep the head private. Not accessible outside this class
 
     # returns the value in the first node
     # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
+        if self.head is None:
+            return None
+        return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n); worst case would be the last element of our list to match our val
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
+        if self.head is None:
+            return False
+        
+        current = self.head
+        while current is not None:
+            if current.value != value:
+                current = current.next
+            else:
+                return True
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n); need to traverse through entire list to get count
+    # Space Complexity: O(1)
     def length(self):

```

Comment:
👍   An alternative solution is that we could maintain a `length` field as a number variable.  Then we can increment it when adding a node and subtract from it when removing a node.